### PR TITLE
Fix Traffic Ops Go to use transactions for deliveryservices, sslkeys, urisigning

### PR DESCRIPTION
**Pull request by rob05c**
The API needs to be using transactions for everything, for multi-query selects and inserts to be correct.

This changes DSes, sslkeys, and urisigning to use transactions.

Sorry about the PR size, it's not really possible to make smaller, changing `db` into `tx` was like pulling a string.

However, this only changes what's necessary for the DS endpoints.

Further, this adds helper functions which automatically start the transaction, so it should be easier to do the other endpoints in smaller increments.

Also adds a validation helper, `api.Parse`, to parse JSON input and validate with a single call.

Tested manually, and API Tests framework passes.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1752/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1759/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1782/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1812/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1828/
Test PASSed.

File: lib/go-tc/deliveryservices.go [hunk 1]
```diff
@@ -205,6 +216,213 @@ type DeliveryServiceNullableV13 struct {
...
+func validateOrgServerFQDN(orgServerFQDN string) bool {
+	_, fqdn, port, err := ParseOrgServerFQDN(orgServerFQDN)
+	if err != nil || !govalidator.IsHost(*fqdn) || (port != nil && !govalidator.IsPort(*port)) {
+		return false
+	}
+	return true
+}
+
+func ParseOrgServerFQDN(orgServerFQDN string) (*string, *string, *string, error) {
+	originRegex := regexp.MustCompile(`^(https?)://([^:]+)(:(\d+))?$`)
```
[the remark this answers is not in the document]
Comment (rob05c) [reply lib/go-tc/deliveryservices.go]: I didn't write this, I just moved it. @rawlinp https://github.com/apache/trafficcontrol/pull/2314

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservicesv12.go [hunk 2]
```diff
@@ -133,35 +96,45 @@ func getDSTenantIDByName(db *sql.DB, name string) (*int, bool, error) {
...
+	return isTenantAuthorized(user, tx, (*tc.DeliveryServiceNullableV12)(ds))
 }
 
-func isTenantAuthorized(user auth.CurrentUser, db *sqlx.DB, ds *tc.DeliveryServiceNullableV12) (bool, error) {
+func isTenantAuthorized(user *auth.CurrentUser, tx *sql.Tx, ds *tc.DeliveryServiceNullableV12) (bool, error) {
 	if ds.ID == nil && ds.XMLID == nil {
 		return false, errors.New("delivery service has no ID or XMLID")
 	}
 
 	existingID, err := (*int)(nil), error(nil)
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/deliveryservice/deliveryservicesv12.go]: Correct, I didn't change it. I agree. In fact, using a single declaration syntax everywhere is even better, ```
existingID := (*int)(nil)
err := error(nil)
```

[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/deliveryservice/deliveryservicesv12.go]: Changed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1835/
Test PASSed.

File: lib/go-tc/deliveryservices.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rawlinp) [reply lib/go-tc/deliveryservices.go]: Yes, that's intentional. The Perl API prohibited IPv6 hosts in the `orgServerFqdn` field but allowed IPv4 hosts, so I just went along with that (not that I think they _should_ be prohibited, but they'd need proper support everywhere first).

**Comment by rob05c** [comment]
@alficles Fixed weasel issues.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1853/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1854/
Test PASSed.

**Pull request merged**